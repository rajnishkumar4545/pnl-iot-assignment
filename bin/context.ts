interface Context {
    referenceTableName: string;
    referenceTablePartitionkey: string;
    vehicleTableName: string;
    vehicleTablePartitionKey: string;
    deviceTableName: string;
    deviceTablePartitionKey: string;
    lambdaLayerName: string;
    lambdaFunctionName: string;
    lambdaAliasName: string;
    snsTopicArn: string;
  }
  
  export const dev: Context = {
    referenceTableName: 'vehicle_device_mapping_table',
    referenceTablePartitionkey: 'device_mac_address',
    vehicleTableName: 'vehicle_table',
    vehicleTablePartitionKey: 'vehicle_mac_address',
    deviceTableName: 'device_table',
    deviceTablePartitionKey: 'device_mac_address',
    lambdaLayerName: 'events-handler-lambda-layer',
    lambdaFunctionName: 'events-handler-lambda-function',
    lambdaAliasName: 'events-handler-lambda-alias',
    snsTopicArn: 'arn:aws:sns:ap-south-1:189351993567:demo-sns-topic',
  };
  
  export const tst: Context = {
    referenceTableName: 'vehicle_device_mapping_table',
    referenceTablePartitionkey: 'device_mac_address',
    vehicleTableName: 'vehicle_table',
    vehicleTablePartitionKey: 'vehicle_mac_address',
    deviceTableName: 'device_table',
    deviceTablePartitionKey: 'device_mac_address',
    lambdaLayerName: 'events-handler-lambda-layer',
    lambdaFunctionName: 'events-handler-lambda-function',
    lambdaAliasName: 'events-handler-lambda-alias',
    snsTopicArn: 'arn:aws:sns:ap-south-1:189351993567:demo-sns-topic',
  };
  