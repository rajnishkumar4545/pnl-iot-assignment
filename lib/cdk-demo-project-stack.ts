import { aws_dynamodb as dynamodb, aws_lambda as lambda ,Stack, StackProps } from 'aws-cdk-lib';
import { DynamoEventSource } from 'aws-cdk-lib/aws-lambda-event-sources';
import { Construct } from 'constructs';
import DynamoDbTable from './constructs/dynamodb';
import DemoPocLambda from './constructs/lambdas';

export class CdkDemoProjectStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);


    // Step 1 : Create DynamoDB Reference Table.
    const referenceTableResource = new DynamoDbTable(this, 'dynamodbRefTable', {
      tableName : 'demo-ref-table',
      partitionKey : 'device_mac_address'
    });

    // Step 2 : Create DynamoDB Vehicle Table.
    const vehicleTableResource = new DynamoDbTable(this, 'dynamodbVehicleTable', {
      tableName : 'demo-vehicle-table',
      partitionKey : 'vehicle_mac_address'
    });


    // Step 3 : Create DynamoDB Reference Table.
    const deviceTableResource = new DynamoDbTable(this, 'dynamodbDeviceTable', {
      tableName : 'demo-device-table',
      partitionKey : 'device_mac_address',
      stream: dynamodb.StreamViewType.NEW_IMAGE
    });


    // Step 5 : create Lambda function
    const lambdaResource = new DemoPocLambda(this, 'pocDemoLambda', {
      dynamodbReferenceTableName : referenceTableResource.getDynamoDbTable().tableName,
      dynamodbVehicleTableName : vehicleTableResource.getDynamoDbTable().tableName,
      dynamodbDeviceTableName : deviceTableResource.getDynamoDbTable().tableName
    });

    
    // // Step 6 : Provide Lambda function permission to read Reference Table
    // referenceTableResource.getDynamoDbTable().grantReadData(lambdaResource.getLambdaFunction());

    // // Step 7 : Provide Lambda function permission to read Vehicle Table
    // vehicleTableResource.getDynamoDbTable().grantReadData(lambdaResource.getLambdaFunction());

    // // Step 8 : Provide Lambda function permission to read stream of Device Table Table
    // deviceTableResource.getDynamoDbTable().grantStreamRead(lambdaResource.getLambdaFunction());

    // // Step 9 : add event source mapping for Device Dynamodb table and Lambda
    // lambdaResource.getLambdaFunction().addEventSource(
    //   new DynamoEventSource(deviceTableResource.getDynamoDbTable(), {
    //     startingPosition: lambda.StartingPosition.LATEST,
    //   })
    // );
  }
}
