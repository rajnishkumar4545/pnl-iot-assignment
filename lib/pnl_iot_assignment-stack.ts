import * as cdk from 'aws-cdk-lib';
import { aws_dynamodb as dynamodb, aws_lambda as lambda, Stack,StackProps } from 'aws-cdk-lib';
import { DynamoEventSource } from 'aws-cdk-lib/aws-lambda-event-sources';
import { Construct } from 'constructs';
import DynamoDbTable from './constructs/dynamodb';
import IotEventsHandlerLambda from './constructs/lambdas';

interface PnlIotAssignmentStackProps extends cdk.StackProps {
  readonly referenceTableName: string;
  readonly referenceTablePartitionkey: string;
  readonly vehicleTableName: string;
  readonly vehicleTablePartitionKey: string;
  readonly deviceTableName: string;
  readonly deviceTablePartitionKey: string;
  readonly lambdaLayerName: string;
  readonly lambdaFunctionName: string;
  readonly lambdaAliasName: string;
  readonly snsTopicArn: string;
}


export class PnlIotAssignmentStack extends Stack {
  constructor(scope: Construct, id: string, props: PnlIotAssignmentStackProps) {
    super(scope, id, props);

    
    // Step 1 : Create DynamoDB Reference Table.
    const referenceTableResource = new DynamoDbTable(this, 'dynamodbRefTable', {
      tableName : props.referenceTableName,
      partitionKey : props.referenceTablePartitionkey
    });

    // Step 2 : Create DynamoDB Vehicle Table.
    const vehicleTableResource = new DynamoDbTable(this, 'dynamodbVehicleTable', {
      tableName :props.vehicleTableName,
      partitionKey : props.vehicleTablePartitionKey
    });


    // Step 3 : Create DynamoDB Reference Table.
    const deviceTableResource = new DynamoDbTable(this, 'dynamodbDeviceTable', {
      tableName : props.deviceTableName,
      partitionKey :props.deviceTablePartitionKey,
      stream: dynamodb.StreamViewType.NEW_IMAGE
    });


    // Step 5 : create Lambda function
    const lambdaResource = new IotEventsHandlerLambda(this, 'pocDemoLambda', {
      dynamodbReferenceTableName : referenceTableResource.getDynamoDbTable().tableName,
      dynamodbVehicleTableName : vehicleTableResource.getDynamoDbTable().tableName,
      dynamodbDeviceTableName : deviceTableResource.getDynamoDbTable().tableName,
      lambdaLayerName : props.lambdaLayerName,
      lambdaFunctionName: props.lambdaFunctionName,
      lambdaAliasName : props.lambdaAliasName,
      snsTopicArn : props.snsTopicArn,
    });

    
    // Step 6 : Provide Lambda function permission to read Reference Table
    referenceTableResource.getDynamoDbTable().grantReadData(lambdaResource.getLambdaFunction());

    // Step 7 : Provide Lambda function permission to read Vehicle Table
    vehicleTableResource.getDynamoDbTable().grantReadData(lambdaResource.getLambdaFunction());

    // Step 8 : Provide Lambda function permission to read stream of Device Table Table
    deviceTableResource.getDynamoDbTable().grantStreamRead(lambdaResource.getLambdaFunction());

    // Step 9 : add event source mapping for Device Dynamodb table and Lambda
    lambdaResource.getLambdaFunction().addEventSource(
      new DynamoEventSource(deviceTableResource.getDynamoDbTable(), {
        startingPosition: lambda.StartingPosition.LATEST,
      })
    );
  }
}
