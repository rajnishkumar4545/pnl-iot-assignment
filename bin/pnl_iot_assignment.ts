#!/usr/bin/env node
import 'source-map-support/register';
import * as cdk from 'aws-cdk-lib';
import { PnlIotAssignmentStack } from '../lib/pnl_iot_assignment-stack';
import * as context from './context';



/*
** Environment variables
*/
const environment = 'dev';
const envContext = context[environment];

const tags = {
  Environment: environment,
  ResourceContact: 'Rajnish Kumar',
};


const app = new cdk.App({
    context: {
      environment,
    },
  });

new PnlIotAssignmentStack(app, 'PnlIotAssignmentStack', {
    referenceTableName: envContext.referenceTableName,
    referenceTablePartitionkey: envContext.referenceTablePartitionkey,
    vehicleTableName: envContext.vehicleTableName,
    vehicleTablePartitionKey: envContext.vehicleTablePartitionKey,
    deviceTableName: envContext.deviceTableName,
    deviceTablePartitionKey: envContext.deviceTablePartitionKey,
    lambdaLayerName: envContext.lambdaLayerName,
    lambdaFunctionName: envContext.lambdaFunctionName,
    lambdaAliasName: envContext.lambdaAliasName,
    snsTopicArn: envContext.snsTopicArn,
});