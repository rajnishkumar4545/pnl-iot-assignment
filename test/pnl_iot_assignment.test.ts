import * as cdk from 'aws-cdk-lib';
import { Capture, Match, Template } from 'aws-cdk-lib/assertions';
import { PnlIotAssignmentStack } from '../lib/pnl_iot_assignment-stack';
import * as context from '../bin/context';

/*
** Environment variables
*/
const environment = 'dev';
const envContext = context[environment];

describe('pnl assessment stack tests', () => {
    test('Should generate stack as expected by snapshot', () => {
    const app = new cdk.App();
      const stack = new PnlIotAssignmentStack(app, 'CdkDemoProjectStack', {
        referenceTableName: envContext.referenceTableName,
    referenceTablePartitionkey: envContext.referenceTablePartitionkey,
    vehicleTableName: envContext.vehicleTableName,
    vehicleTablePartitionKey: envContext.vehicleTablePartitionKey,
    deviceTableName: envContext.deviceTableName,
    deviceTablePartitionKey: envContext.deviceTablePartitionKey,
    lambdaLayerName: envContext.lambdaLayerName,
    lambdaFunctionName: envContext.lambdaFunctionName,
    lambdaAliasName: envContext.lambdaAliasName,
    snsTopicArn: envContext.snsTopicArn,
      });
      const template = Template.fromStack(stack);

      // Check for Lambda Property
      template.hasResourceProperties("AWS::Lambda::Function", {
        Handler: "iot_events_handler.lambda_handler",
        Runtime: "python3.9",
        Timeout : 60,
        FunctionName: "events-handler-lambda-function",
        Architectures : [ "arm64" ],
        TracingConfig : { "Mode": "Active" },
        Environment : {
          "Variables": {
            "REFERENCE_TABLE_NAME": {
             "Ref": "dynamodbRefTablevehicledevicemappingtabledynamodbLogicalIdFA79C9F6"
            },
            "VEHICLE_TABLE_NAME": {
             "Ref": "dynamodbVehicleTablevehicletabledynamodbLogicalId917C0593"
            },
            "DEVICE_TABLE_NAME": {
             "Ref": "dynamodbDeviceTabledevicetabledynamodbLogicalIdE02DFAD2"
            }
           }
        }
      });

      template.hasResourceProperties("AWS::DynamoDB::Table", {
        TableClass : "STANDARD",
        TableName : "vehicle_device_mapping_table",
        BillingMode: "PAY_PER_REQUEST",
        SSESpecification : { "SSEEnabled": true },
        KeySchema : [ 
          {
            "AttributeName": "device_mac_address",
            "KeyType": "HASH"
           }
          ],
          "AttributeDefinitions": [
           {
            "AttributeName": "device_mac_address",
            "AttributeType": "S"
           }
         ]

      });

      // Test for Number of Resource Created
      template.resourceCountIs("AWS::DynamoDB::Table", 3);
      template.resourceCountIs("AWS::IAM::Role", 2);
      template.resourceCountIs("Custom::LogRetention", 1);
      template.resourceCountIs("AWS::IAM::Policy", 2);
      template.resourceCountIs("AWS::Lambda::Function", 2);

      // Sanpshot test
      expect(template.toJSON()).toMatchSnapshot();
    });
  });
