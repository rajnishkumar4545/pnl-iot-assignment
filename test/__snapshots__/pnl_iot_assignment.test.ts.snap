// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pnl assessment stack tests Should generate stack as expected by snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "eb5b005c858404ea0c8f68098ed5dcdf5340e02461f149751d10f59c210d5ef8.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "dynamodbDeviceTabledevicetabledynamodbLogicalIdE02DFAD2": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "device_mac_address",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "device_mac_address",
            "KeyType": "HASH",
          },
        ],
        "SSESpecification": {
          "SSEEnabled": true,
        },
        "StreamSpecification": {
          "StreamViewType": "NEW_IMAGE",
        },
        "TableClass": "STANDARD",
        "TableName": "device_table",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "dynamodbRefTablevehicledevicemappingtabledynamodbLogicalIdFA79C9F6": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "device_mac_address",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "device_mac_address",
            "KeyType": "HASH",
          },
        ],
        "SSESpecification": {
          "SSEEnabled": true,
        },
        "TableClass": "STANDARD",
        "TableName": "vehicle_device_mapping_table",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "dynamodbVehicleTablevehicletabledynamodbLogicalId917C0593": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "vehicle_mac_address",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "vehicle_mac_address",
            "KeyType": "HASH",
          },
        ],
        "SSESpecification": {
          "SSEEnabled": true,
        },
        "TableClass": "STANDARD",
        "TableName": "vehicle_table",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "eventsHandlerAlias18D81974": {
      "Properties": {
        "FunctionName": {
          "Ref": "pocDemoLambdaEventsHandlerLambda62388BA9",
        },
        "FunctionVersion": {
          "Fn::GetAtt": [
            "eventsHandlerLambdaVersionIdF2189976",
            "Version",
          ],
        },
        "Name": "events-handler-lambda-alias",
      },
      "Type": "AWS::Lambda::Alias",
    },
    "eventsHandlerLambdaVersionIdF2189976": {
      "Properties": {
        "FunctionName": {
          "Ref": "pocDemoLambdaEventsHandlerLambda62388BA9",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "pocDemoLambdaEventsHandlerLambda62388BA9": {
      "DependsOn": [
        "pocDemoLambdaEventsHandlerLambdaServiceRoleDefaultPolicy70C38AE7",
        "pocDemoLambdaEventsHandlerLambdaServiceRoleDDA5FF74",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "0b82de6492c08a362eb732440533a8229c5afc784a3d4f40d96c20838599c42d.zip",
        },
        "Description": "IoT Events Handler Lambda Function",
        "Environment": {
          "Variables": {
            "DEVICE_TABLE_NAME": {
              "Ref": "dynamodbDeviceTabledevicetabledynamodbLogicalIdE02DFAD2",
            },
            "REFERENCE_TABLE_NAME": {
              "Ref": "dynamodbRefTablevehicledevicemappingtabledynamodbLogicalIdFA79C9F6",
            },
            "SNS_ARN": "arn:aws:sns:ap-south-1:189351993567:demo-sns-topic",
            "VEHICLE_TABLE_NAME": {
              "Ref": "dynamodbVehicleTablevehicletabledynamodbLogicalId917C0593",
            },
          },
        },
        "FunctionName": "events-handler-lambda-function",
        "Handler": "iot_events_handler.lambda_handler",
        "Layers": [
          {
            "Ref": "pocDemoLambdaeventsHandlerLambdaLayer01C20414",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "pocDemoLambdaEventsHandlerLambdaServiceRoleDDA5FF74",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 60,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "pocDemoLambdaEventsHandlerLambdaDynamoDBEventSourceCdkDemoProjectStackdynamodbDeviceTabledevicetabledynamodbLogicalIdD4C4D88CCDD714C1": {
      "Properties": {
        "BatchSize": 100,
        "EventSourceArn": {
          "Fn::GetAtt": [
            "dynamodbDeviceTabledevicetabledynamodbLogicalIdE02DFAD2",
            "StreamArn",
          ],
        },
        "FunctionName": {
          "Ref": "pocDemoLambdaEventsHandlerLambda62388BA9",
        },
        "StartingPosition": "LATEST",
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "pocDemoLambdaEventsHandlerLambdaLogRetention4B2A9A81": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "pocDemoLambdaEventsHandlerLambda62388BA9",
              },
            ],
          ],
        },
        "RetentionInDays": 7,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "pocDemoLambdaEventsHandlerLambdaServiceRoleDDA5FF74": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "pocDemoLambdaEventsHandlerLambdaServiceRoleDefaultPolicy70C38AE7": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "logs:CreateLogGroup",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:logs:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:logs:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group:/aws/lambda/*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "sns:Publish",
                "sns:ListTopics",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:sns:ap-south-1:189351993567:demo-sns-topic",
            },
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "dynamodbRefTablevehicledevicemappingtabledynamodbLogicalIdFA79C9F6",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "dynamodbVehicleTablevehicletabledynamodbLogicalId917C0593",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
            {
              "Action": "dynamodb:ListStreams",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "dynamodb:DescribeStream",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "dynamodbDeviceTabledevicetabledynamodbLogicalIdE02DFAD2",
                  "StreamArn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "pocDemoLambdaEventsHandlerLambdaServiceRoleDefaultPolicy70C38AE7",
        "Roles": [
          {
            "Ref": "pocDemoLambdaEventsHandlerLambdaServiceRoleDDA5FF74",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "pocDemoLambdaeventsHandlerLambdaLayer01C20414": {
      "Properties": {
        "CompatibleArchitectures": [
          "arm64",
        ],
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "a966740851c2f8630962d4b5ca81f388ab2e0fb74b57a141d42e6808f8b3cebe.zip",
        },
        "LayerName": "events-handler-lambda-layer",
        "LicenseInfo": "Assignment Purpose",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
